graspologic/embed/case.py

```diff
@@ -0,0 +1,209 @@
+        self.is_fitted_ = False
+
+    def fit(
+        self, network: Tuple[np.ndarray, np.ndarray], y: None = None
+    ) -> "CovariateAssistedEmbed":
+        """
+        Fit a CASE model to an input graph, along with its covariates. Depending on the
+        embedding algorithm, we embed
+
+        .. math:: L_ = LL + \alpha YY^T
+        .. math:: L_ = L + \alpha YY^T
+
+        where :math:`\alpha` is a tuning parameter which makes the leading eigenvalues
+        of the two summands the same. Here, :math:`L` is the regularized
+        graph Laplacian, and :math:`Y` is a matrix of covariates for each node.
+
+        Covariates are row-normalized to unit l2-norm.
+
+        Parameters
+        ----------
+        network : tuple or list of np.ndarrays
+            Contains the tuple (A, Y), where A is an adjacency matrix and Y is the
+            matrix of covariates.
+
+            A : array-like or networkx.Graph
+                Input graph to embed. See graspologic.utils.import_graph
+
+            Y : array-like, shape (n_vertices, n_covariates)
+                Covariate matrix. Each node of the graph is associated with a set of
+                `d` covariates. Row `i` of the covariates matrix corresponds to node
+                `i`, and the number of columns are the number of covariates.
+
+        y: Ignored
+
+        Returns
+        -------
+        self : object
+            Returns an instance of self.
+        """
+
+        # setup
+        if not isinstance(network, (tuple, list)):
+            msg = "Network should be a tuple-like object of (graph, covariates)."
+            raise TypeError(msg)
+        if len(network) != 2:
+            msg = "Network should be a tuple-like object of (graph, covariates)."
+            raise ValueError(msg)
+
+        graph, covariates = network
+        A = import_graph(graph)
+
+        # Create regularized Laplacian, scale covariates to unit norm
+        L = to_laplacian(A, form="R-DAD")
+        Y = covariates
+        if Y.ndim == 1:
+            Y = Y[:, np.newaxis]
+        Y = normalize(Y, axis=0)
+
+        # Use ratio of the two leading eigenvalues if alpha is None
+        self._get_tuning_parameter(L, Y)
+
+        # get embedding matrix as a LinearOperator (for computational efficiency)
+        n = A.shape[0]
+        mv, rmv = self._matvec(L, Y, a=self.alpha_, assortative=self.assortative)
+        L_ = LinearOperator((n, n), matvec=mv, rmatvec=rmv)
+
+        # Dimensionality reduction with SVD
+        self._reduce_dim(L_)
+
+        self.is_fitted_ = True
+        return self
+
+    def _get_tuning_parameter(
+        self, L: np.ndarray, Y: np.ndarray
+    ) -> "CovariateAssistedEmbed":
+        """
+        Find the alpha which causes the leading eigenspace of LL and YYt to be the same.
+
+        Parameters
+        ----------
+        L : array
+            The regularized graph Laplacian.
+        Y : array
+            The covariate matrix.
+
+        Returns
```

Comment:
I believe this should be an attribute, and the return should be self

Reply:
yep, good catch, d8740b5